### ISSUE TITLE
Deterministic behaviour with torch.randn_like() on mps when tensor dimensionality exceeds some size

### ISSUE THREAD

**Issue by henry-ald** [pytorch/pytorch#147740]
### 🐛 Describe the bug

With `device="mps"`, `torch.randn_like()` is producing tensors with elements of identical value along a given dimension, specifically once the dimensionality exceeds a certain size. This behaviour is not present on the CPU. Here is some code demonstrating:

```import torch

# On MPS GPU
X = torch.randn(4, device="mps", dtype=torch.float32)
print(torch.randn_like(X))

X = torch.randn(4,1, device="mps", dtype=torch.float32)
print(torch.randn_like(X))

X = torch.randn(4,1,1, device="mps", dtype=torch.float32)
print(torch.randn_like(X))

X = torch.randn(4,1,1,1, device="mps", dtype=torch.float32)
print(torch.randn_like(X))

X = torch.randn(4,1,1,1,1, device="mps", dtype=torch.float32)
print(torch.randn_like(X)) # Elements of identical value

X = torch.randn(4,1,1,1,1,1, device="mps", dtype=torch.float32)
print(torch.randn_like(X)) # Elements of identical value


# On CPU
X = torch.randn(4)
print(torch.randn_like(X))

X = torch.randn(4,1)
print(torch.randn_like(X))

X = torch.randn(4,1,1)
print(torch.randn_like(X))

X = torch.randn(4,1,1,1)
print(torch.randn_like(X))

X = torch.randn(4,1,1,1,1)
print(torch.randn_like(X)) # Elements NOT of identical value

X = torch.randn(4,1,1,1,1,1)
print(torch.randn_like(X)) # Elements NOT of identical value
```

For example, `X = torch.randn(4,1,1,1,1, device="mps", dtype=torch.float32)` produces the tensor 

```
tensor([[[[[-0.2945]]]],



        [[[[-0.2945]]]],



        [[[[-0.2945]]]],



        [[[[-0.2945]]]]], device='mps:0')
```

 while its CPU counterpart `X = torch.randn(4,1,1,1,1)` will produce 

```
tensor([[[[[ 0.6925]]]],



        [[[[ 2.0277]]]],



        [[[[-0.0787]]]],



        [[[[ 0.0240]]]]])
```

This deterministic behaviour does not occur for the examples of tensor size less than or equal to (4,1,1,1).

### Versions

```
Collecting environment information...
PyTorch version: 2.5.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.3 (arm64)
GCC version: Could not collect
Clang version: 16.0.0 (clang-1600.0.26.6)
CMake version: Could not collect
Libc version: N/A

Python version: 3.13.2 | packaged by Anaconda, Inc. | (main, Feb  6 2025, 12:55:35) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-15.3-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] numpy==2.2.2
[pip3] torch==2.5.1
[conda] libtorch                  2.5.1           gpu_mps_h82d5d13_202  
[conda] nomkl                     3.0                           0  
[conda] numpy                     2.2.2           py313h7c57ca2_0  
[conda] numpy-base                2.2.2           py313hb98e858_0  
[conda] pytorch                   2.5.1           gpu_mps_py313h80af30b_202  
```

**Comment by malfet** [comment]
This looks like a duplicate of https://github.com/pytorch/pytorch/issues/147624

**Comment by malfet** [comment]
@henry-ald no worries. Can you please try latest nightly to make sure that problem is gone (building from source now)

**Comment by henry-ald** [comment]
@malfet It is still producing the same error, and I believe I installed the nightly build from source correctly:

```
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: macOS 15.3 (arm64)
GCC version: Could not collect
Clang version: 16.0.0 (clang-1600.0.26.6)
CMake version: version 3.31.4
Libc version: N/A

Python version: 3.13.2 | packaged by Anaconda, Inc. | (main, Feb  6 2025, 12:55:35) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-15.3-arm64-arm-64bit-Mach-O
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] numpy==2.2.3
[pip3] optree==0.14.0
[pip3] torch==2.7.0a0+git576ed1e
[conda] numpy                     2.2.3                    pypi_0    pypi
[conda] optree                    0.14.0                   pypi_0    pypi
[conda] torch                     2.7.0a0+git576ed1e           dev_0    <develop>
```

I'm working on some research software in a private repo, not following any tutorials. I agree it is weird that we both notice it just now, I suppose it's just chance.

**Comment by malfet** [comment]
Adding needs reproduction label, as I can not repro the results after https://github.com/pytorch/pytorch/pull/147667

**Comment by jhavukainen** [comment]
Unable to reproduce the issue after the workaround on either. Filed an issue against MPSGraph based on the reproducer from https://github.com/pytorch/pytorch/pull/147667 (thanks @malfet) to get them to also address the underlying issue.

**Comment by malfet** [comment]
Closing, as all random ops has been migrated to Metal and seems to be working fine
